<nav class="navbar navbar-default">
  <div class="container-fluid">
    <div class="navbar-header">
      <a class="navbar-brand" href="_Dataoving2.ipynb">Dataøving 2</a>
    </div>
    <ul class="nav navbar-nav">
        <li><a href="Oppgave1.ipynb">Oppgave1 - Digitale Signal og Diskrét Datarepresentasjon.</a></li>
        <li class="active"><a href="Oppgave2.ipynb">Oppgave 2 - Signalanalyse med DFT.</a></li>
        <li><a href="Oppgave3.ipynb">Oppgave 3 - Signalgenerasjon med IDFT.</a></li>
        <li><a href="Oppgave4.ipynb">Oppgave 4 - Reparasjon av lydsignal.</a></li>
    </ul>
  </div>
</nav>

# Signalanalyse med DFT

__Læringsmål__:
* Forståelse for sammenheng mellom DFT-indeks og frekvens.
* Tolking av Amplituden til sinuskomponenter i en N-punkts DFT.
* Tolking av Fasen til sinuskomponenter i en N-punkts DFT.

__Bibliotek og notebook-konfigurasjon:__

In [1]:
from numpy import sin, cos, pi, exp, append, mean, absolute, angle # Sentrale matematiske funksjoner
from numpy.fft import fft, ifft             # DFT og IDFT
import numpy as np                          # Importer funksjonalitet fra numpy biblioteket med prefiks "np"
import matplotlib.pyplot as plt             # Importer pyplot modulen i matplotlib med prefiks "plt"
import csv                                  # Håndtering av .csv fil
from funksjoner_data3 import timeSpectrumStem
from scipy.io import loadmat

%matplotlib ipympl

**Introduksjon**

I denne oppgaven skal vi se på et testsignal $x[n]$ med 128 sampler, bestående av to sinussekvenser som er lagt sammen $x[n] = A_1\cdot \cos\left(\hat{\omega}_1\cdot n + \phi_1 \right) + A_2\cdot \cos\left(\hat{\omega}_2\cdot n + \phi_2 \right)$. Signalet lastes inn fra datafila `lab3.mat`, og er signalet gitt i kolonnen merket `x1[n]`. Ved å regne DFT (fft) av dette signalet kan vi analysere frekvensinnholdet i signalet. Koden nedenfor vil laste inn signalet `x1[n]` fra datafila, regne ut signalets DFT, og visualisere resultatet ved bruk av funksjonen `timeSpectrumStem()` som er laget for denne dataøvingen. Denne funksjonen tar in digitalt signal som tidsserie (`xn1`), signalets DFT (`Xk1`), og antall sampler (`N`) som parametre.

In [2]:
fileData = loadmat('lab3.mat', squeeze_me=True)

n = fileData['n']
xn1 = fileData['x1[n]']

Xk1 = fft(xn1);
N = len(xn1);

plt.close(1); 
plt.figure(1);
timeSpectrumStem(xn1, Xk1, N)

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

## a)

Studer figurene. For hvilke indeksverdier $k$ finner vi elementene i DFT-sekvensen $X[k]$ som hører til de to sinuskomponentene med frekvens $\hat{\omega}_1$ og $\hat{\omega}_2$? 

$\hat{\omega}_1$:
* Ett komplekskonjugert "par" nær starten og slutten av DFT-sekvensen $X[k]$ 
    - Positiv frekvenskomponent: $k=2$ 
    - Alias av negativ frekvenskomponent: $k=N-2=128-2=126$
    
    
$\hat{\omega}_2$:
* Ett komplekskonjugert "par" nærmere midten av DFT-sekvensen $X[k]$
    - Positiv frekvenskomponent: $k=33$
    - Alias av negativ frekvenskomponent: $k=N-33=128-33=95$

## b)
Vi kan skrive ut numeriske verdier til utvalgte element i $X_1[k]$, og ett "par" med symmetrisk plasserte element representerer én sinusbølge. Ett eksempel på hvordan man kan finne verdien til eksempelvis $X_1[81]$ i polarkoordinater er vist nedenfor:
```python
k = 81
print("X[%d] har absoluttverdi %.1f og vinkel %.2f Pi"% (k, absolute(Xk1[k]), angle(Xk1[k])/pi))
```
- Studer de ulike sinuskomponentene til $x_1[n]$ i frekvensplanet ved å skrive ut verdien(e) til alle DFT-elementene som tilhører de to sinusbølgene som utgjør signalet. Finn så amplitude $A$, normalisert vinkelfrekvens $\hat{\omega}$ og fase $\phi$ for hver sinuskomponent.
- Hvis signalet har en samplingsfrekvens $f_s = 4096Hz$, hva er frekvensen til de to sinuskomponentene $f_1$ og $f_2$ i Hz?

In [3]:
k = [2, 33, 95, 126]      # DFT Element 
for i in range(len(k)):
    print("X[%d] har absoluttverdi %.1f og vinkel %.2f Pi"% (k[i], absolute(Xk1[k[i]]), angle(Xk1[k[i]])/pi))

X[2] har absoluttverdi 64.0 og vinkel -0.50 Pi
X[33] har absoluttverdi 25.6 og vinkel -0.75 Pi
X[95] har absoluttverdi 25.6 og vinkel 0.75 Pi
X[126] har absoluttverdi 64.0 og vinkel 0.50 Pi


**Komponent 1:**

- $f_1 = \frac{2}{N}\cdot f_s = \frac{2}{128}\cdot 4096 = 64Hz$
- $A_1 = \frac{|X[2]|+|X[126|}{N}=\frac{64+64}{128}=1$
- $\phi_1 = \angle{X[2]} = -\frac{\pi}{2}$

**Komponent 2:**

- $f_2 = \frac{33}{N}\cdot f_s = \frac{33}{128}\cdot 4096 = 1056Hz$
- $A_2 = \frac{|X[33]|+|X[95|}{N}=\frac{25.6+25.6}{128}=0.4$
- $\phi_2 = \angle{X[33]} = -\frac{3\pi}{4}$